### ISSUE TITLE
Bootstrap.py fails in version 4.x

### ISSUE THREAD

**Issue by texadactyl** [spyder-ide/spyder#13784]
## Problem Description

While trying to install PyLS locally, bootstrap.py is unable to create the required directory spyder-4.x/external-deps/python-language-server/.installation-dir.

### What steps reproduce the problem?

I followed the instructions "To start Spyder directly from your clone": https://github.com/spyder-ide/spyder/blob/master/CONTRIBUTING.md

### What is the expected output? What do you see instead?

Expected: Spyder launched.
Observed: Traceback.

### Paste Traceback/Error Below (if applicable)

```python-traceback

python3 bootstrap.py
Executing Spyder from source checkout
WARNING: Ubuntu Apport exception hook is detected
         Use --no-apport option to disable it
*. Patched sys.path with /home/elkins/Downloads/spyder
*. Patched sys.path with /home/elkins/Downloads/spyder/external-deps/python-language-server
*. Patched sys.path with /home/elkins/Downloads/spyder/external-deps/spyder-kernels
*. Installing PyLS locally
error: can't create or remove files in install directory

The following error occurred while trying to add or remove files in the
installation directory:

    [Errno 2] No such file or directory: '/home/elkins/Downloads/spyder/external-deps/python-language-server/.installation-dir/test-easy-install-216349.write-test'

```

## Versions

* Spyder version: 4.x from github
* Python version: 3.8.2
* Qt version: 5.12.8
* PyQt version: 5.14.1
* Operating System name/version: Xubuntu 20.04.up-to-date

### Fix
Replace line 165:
```
print("*. Installing PyLS locally")
```
with
```
print("*. Installing PyLS locally on {}".format(pyls_installation_dir))
if not osp.exists(pyls_installation_dir):
    Path(pyls_installation_dir).mkdir(parents=True)
```

Insert the following after line 26:
```
from pathlib import Path
```

**Comment by texadactyl** [comment]
Almost forgot the log of success.  From stdout:
```
python3 bootstrap.py
Executing Spyder from source checkout
WARNING: Ubuntu Apport exception hook is detected
         Use --no-apport option to disable it
*. Patched sys.path with /home/elkins/Downloads/spyder-4.x
*. Patched sys.path with /home/elkins/Downloads/spyder-4.x/external-deps/python-language-server
*. Patched sys.path with /home/elkins/Downloads/spyder-4.x/external-deps/spyder-kernels
*. Installing PyLS locally on /home/elkins/Downloads/spyder-4.x/external-deps/python-language-server/.installation-dir
*. PyQt5 is detected, selecting
*. Imported Spyder 4.2.0.dev0 - Revision None, Branch: None
    [Python 3.8.2 64bits, Qt 5.12.9, PyQt5 5.12.3 on Linux]
*. Running Spyder
Bootstrap completed in 00:00:02.0628

```

**Comment by steff456** [comment]
Hi @texadactyl,

Can you check that you have the right permissions on the folder?

**Comment by texadactyl** [comment]
The folder does not exist.
Precisely what I did:
- Downloaded ZIP file.
- Unzipped.
- cd into spyder folder.
- python3 bootstrap.py

Made the indicated changes in bootstrap.py and it works fine.

**Comment by texadactyl** [comment]
If I had a permissions issue, it would not have run no matter what I edited.

**Comment by texadactyl** [comment]
If you are in folder A and you need to create A.B.C, Path.mkdir will not only create C but also B  too.

**Comment by texadactyl** [comment]
It might be useful if the _pyls_installation_dir_ variable could be overwritten via a argparse.ArgumentParser optional argument.  That could be an alternate solution.

**Comment by ccordoba12** [comment]
> The folder does not exist.

What folder doesn't exist?

> Precisely what I did

According to our [contributing guide](https://github.com/spyder-ide/spyder/blob/master/CONTRIBUTING.md#setting-up-a-development-environment), you're expected to run `git clone`. And when you do that, things work just fine.

**Comment by texadactyl** [comment]
spyder-4.x/external-deps/python-language-server/.installation-dir

**Comment by texadactyl** [comment]
I tried to clone 4.x but this happened:
git clone https://github.com/spyder-ide/spyder/tree/4.x
Cloning into '4.x'...
fatal: repository 'https://github.com/spyder-ide/spyder/tree/4.x/' not found

So, I just went into https://github.com/spyder-ide/spyder/tree/4.x, downloaded the ZIP file, and unzipped.

**Comment by texadactyl** [comment]
Ok, this is the right way to clone 4.x:
git clone https://github.com/spyder-ide/spyder -b 4.x

**Comment by texadactyl** [comment]
But, it still fails until one fixes bootstrap.py in some manner (my suggestion or other):
```
git clone https://github.com/spyder-ide/spyder -b 4.x
Cloning into 'spyder'...
remote: Enumerating objects: 120, done.
remote: Counting objects: 100% (120/120), done.
remote: Compressing objects: 100% (96/96), done.
remote: Total 135883 (delta 66), reused 42 (delta 21), pack-reused 135763
Receiving objects: 100% (135883/135883), 74.54 MiB | 2.33 MiB/s, done.
Resolving deltas: 100% (103742/103742), done.


cd spyder
elkins@biostar:~/Downloads/spyder$ python3 bootstrap.py
Executing Spyder from source checkout
WARNING: Ubuntu Apport exception hook is detected
         Use --no-apport option to disable it
*. Patched sys.path with /home/elkins/Downloads/spyder
*. Patched sys.path with /home/elkins/Downloads/spyder/external-deps/python-language-server
*. Patched sys.path with /home/elkins/Downloads/spyder/external-deps/spyder-kernels
*. Installing PyLS locally
error: can't create or remove files in install directory

The following error occurred while trying to add or remove files in the
installation directory:

    [Errno 2] No such file or directory: '/home/elkins/Downloads/spyder/external-deps/python-language-server/.installation-dir/test-easy-install-219572.write-test'

The installation directory you specified (via --install-dir, --prefix, or
the distutils default setting) was:

    /home/elkins/Downloads/spyder/external-deps/python-language-server/.installation-dir

This directory does not currently exist.  Please create it and try again, or
choose a different installation directory (using the -d or --install-dir
option).

Traceback (most recent call last):
  File "bootstrap.py", line 166, in <module>
    subprocess.check_output(
  File "/usr/lib/python3.8/subprocess.py", line 411, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/usr/lib/python3.8/subprocess.py", line 512, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/usr/bin/python3', '-W', 'ignore', 'setup.py', 'develop', '--no-deps', '--install-dir', '/home/elkins/Downloads/spyder/external-deps/python-language-server/.installation-dir']' returned non-zero exit status 1.

```

**Comment by steff456** [comment]
Did you prepare a conda environment as the contributing guide suggests?

**Comment by texadactyl** [comment]
No, I did not.  Unfortunately, the SETI Institute Breakthrough Listen department does not use Anaconda.  Most of my Python work lately has been with them.  Hence, I have been using python3-pip which seems to be in good health.  I have done maintenance in pypi/pip before so I can handle this environment.

This issue is not critical for me.  Feel free to prioritize other issues and enhancements.  I was simply curious about how the IDE was constructed.  I found the developer documentation which is a good roadmap.

As I suspected, whether one downloads the zip and unzips or git-clones, it makes no difference.  The result is the same directory tree.  Except for the bootstrap.py issue and issue #13347, everything is working fine.

**Comment by ccordoba12** [comment]
I think the problem here has to with either the lack of `setuptools` in your environment (which is needed to run `python setup.py develop`) or an old version of it. I'm mentioning that because `pyls_installation_dir` is created automatically in a conda env and it's removed when you exit `bootstrap.py`.

In any case, your solution is not the right one and (unfortunately) we don't have time to take a look at this because things are working fine in our recommended setup (i.e. a conda env), sorry.

**Comment by ccordoba12** [comment]
> Except for the bootstrap.py issue and issue #13347, everything is working fine.

Still no fix for that one, sorry too. But I plan to work on it next week.

**Comment by texadactyl** [comment]
My setuptools: 45.2.0.
I always keep all of the pypi-based packages up-to-date, just like my xubuntu 20.04.

I appreciate what you're saying about using Anaconda.  I am encouraging scientist users to migrate desktops ASAP.  Of course, servers and github/Travis processes will stay pip-oriented.

Not trying to put a load on you folks.  I already went back to the latest spyder.  I can wait on #13347.

**Comment by texadactyl** [comment]
One tactic that seems to work:
(1) Install and use Anaconda as much as possible.
(2) For packages that are not Anaconda-ready, operate in a separate virtual machine.  Don't mix with the local Anaconda repository.

**Comment by ccordoba12** [comment]
> Not trying to put a load on you folks. I already went back to the latest spyder. I can wait on #13347.

Thanks for understanding.

**Comment by CAM-Gerlach** [comment]
@ccordoba12 I can reproduce the exact same error message (with some additional information) on Windows in my fully conda-installed Spyder dev environment (originally created via the steps in the contributing guide perhaps a few months back) when I run `python bootstrap.py` in my Spyder dev environment with PR #13281 checked out, after not having run it from some weeks (before you bumped the PyLS version recently, as I recall). It occurs in both Git Bash and Anaconda Prompt, and persisted after I re-ran ``conda install -c spyder-ide --file requirements/conda.txt`` to update deps to the latest pinned versions.

What finally fixed it was upgrading python, pip, wheel, setuptools and conda to their latest versions in `base` and my Spyder dev env and re-running bootstrap (though there are still PyLS compat issues that I presume are unrelated). If `python-language-server` needs a specific `setuptools` version or similar, at least to install in dev mode (what I assume is happening here), then it should be specified in `requirements/conda.txt`. Do you know what package and version spec fixes this, or would you like me to research/test this?

Full message/traceback:

<details>

$ python bootstrap.py
Executing Spyder from source checkout
*. Patched sys.path with C:\Users\C. A. M. Gerlach\Documents\dev\SpyderDev\spyder
*. Patched sys.path with C:\Users\C. A. M. Gerlach\Documents\dev\SpyderDev\spyder\external-deps\python-language-server

*. Patched sys.path with C:\Users\C. A. M. Gerlach\Documents\dev\SpyderDev\spyder\external-deps\spyder-kernels
*. Installing PyLS locally
error: can't create or remove files in install directory

The following error occurred while trying to add or remove files in the
installation directory:

    [Errno 2] No such file or directory: 'C:\\Users\\C. A. M. Gerlach\\Documents\\dev\\SpyderDev\\spyder\\external-deps\\python-language-server\\.installation-dir\\test-easy-install-11680.write-test'

The installation directory you specified (via --install-dir, --prefix, or
the distutils default setting) was:

    C:\Users\C. A. M. Gerlach\Documents\dev\SpyderDev\spyder\external-deps\python-language-server\.installation-dir

This directory does not currently exist.  Please create it and try again, or
choose a different installation directory (using the -d or --install-dir
option).

Traceback (most recent call last):
  File "bootstrap.py", line 176, in <module>
    cwd=pyls_submodule
  File "C:\Miniconda3\envs\spyder-test\lib\subprocess.py", line 411, in check_output
    **kwargs).stdout
  File "C:\Miniconda3\envs\spyder-test\lib\subprocess.py", line 512, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['C:\\Miniconda3\\envs\\spyder-test\\python.exe', '-W', 'ignore', 'setup.py', 'develop', '--no-deps', '--install-dir', 'C:\\Users\\C. A. M. Gerlach\\Documents\\dev\\SpyderDev\\spyder\\external-deps\\python-language-server\\.installation-dir']' returned non-zero exit status 1.

</details>

**Comment by ccordoba12** [comment]
> Do you know what package and version spec fixes this, or would you like me to research/test this?

I'd really appreciate your help with this one.

I think @dalthviz also mentioned at some point that doing `git clean -xfdi` fixes this problem (perhaps because it removes the previous way we had to install the PyLS in development mode).

**Comment by texadactyl** [comment]
@CAM-Gerlach 
Does my suggested fix solve your issue with an Anaconda environment on Windows?  I would guess so.

**Comment by CAM-Gerlach** [comment]
@ccordoba12 @texadactyl I can give both a shot, assuming I can reliably repro it. Your fix looks fine, apart from some semantic improvements. If a later version of `setuptools` fixes this, it seems it lies between (45.2.0, 49.6.0] since I had an older version than you previously (42.x).

**Comment by real-yfprojects** [comment]
I had the same issue on a `venv` virtualenvironment. But upgrading `setuptools` from 44.0.0 to 53.1.0 fixed it.
